### PR TITLE
feat(sdk): add client side compression for high-level rust API

### LINKED ISSUE #30: Client-side message compression feature
This feature would add optional client-side compression to reduce network bandwidth between clients and servers.

Proposed Implementation

1. Use custom message headers to indicate compression (see field `user_headers` in `IggyMessage`):
```rust
const COMPRESSION_HEADER_KEY: &str = "iggy-compression";  // Values: "gzip", "zstd", "lz4", "snappy"
```

2. Add compression config to producer:
```rust
  pub struct ClientCompressionConfig {
      pub algorithm: ClientCompressionAlgorithm,  // Gzip, Zstd, Lz4, Snappy
      pub min_size: u32,  // Only compress if payload > min_size (default: 1KB)
  }
```
3. Producer compresses paylaod before sending, adds custom user header to `IggyMessage` (K:V)
4. Consumer detects header and decompresses transparently

Tasks

- Add compression deps to Cargo.toml (flate2, zstd, lz4, snap) with feature flags
- Implement compress/decompress in producer/consumer message flow
- Add tests for different algorithms and edge cases
- Benchmark bandwidth vs CPU trade-offs using `iggy-bench` (also implement handling of compression for it)

### PR DESCRIPTION
**Pull request by haubur**
# Implementing Client Side (De-)compression
Closes #30 

- adds `ClientCompressionConfig` based on `CompressionAlgorithm` with min_size
- client producer can be set with `compressor` aka `ClientCompressionConfig`
- compression type is added to `user-headers` in `IggyMessage`
- consumer checks for `iggy-compression` user header and decompresses if present
- adds unit tests for compression algorithms
- adds integration test for compression

@hubcio opened as Draft PR. If implementation is fine, I would extend iggy-bench accordingly and run benchmarks.

### CODE REVIEW

File: core/sdk/src/clients/consumer.rs [hunk 1]
```diff
@@ -980,6 +977,48 @@ impl Stream for IggyConsumer {
...
                                 message.header.payload_length = message.payload.len() as u32;
                             }
+
+                            // maybe decompress
+                            if let Ok(Some(algorithm_value)) = message
+                                .get_user_header(&HeaderKey::from_str("iggy-compression").unwrap())
+                            {
+                                let algorithm_name = algorithm_value.as_str().unwrap();
+                                let algorithm =
+                                    CompressionAlgorithm::from_str(algorithm_name).unwrap();
```
Comment (hubcio) [inline core/sdk/src/clients/consumer.rs]: implement `TryFromStr` for CompressionAlgorithm, dont unwrap

EDIT: i see that it implements `FromStr`, so just stop using unwrap :)

File: core/common/src/types/compression/compression_algorithm.rs [hunk 2]
```diff
@@ -198,4 +394,57 @@ mod tests {
         let invalid_compression_kind = CompressionAlgorithm::from_code(255);
         assert!(invalid_compression_kind.is_err());
     }
+
+    #[test]
+    fn test_emtpy_input_compression_decompression() {
```
Comment (hubcio) [inline core/common/src/types/compression/compression_algorithm.rs]: typo

File: core/common/src/types/compression/compression_algorithm.rs [hunk 3]
```diff
@@ -45,24 +49,110 @@ impl FromStr for CompressionAlgorithm {
...
+            CompressionAlgorithm::Snappy => {
+                let compressed_data = snap::raw::Encoder::new()
+                    .compress_vec(data)
+                    .map_err(|e| IggyError::CompressionError(e.to_string()))?;
+                Ok(compressed_data)
+            }
+        }
+    }
+
+    pub fn decompress(&self, data: &[u8]) -> Result<Vec<u8>, IggyError> {
```
Comment (hubcio) [inline core/common/src/types/compression/compression_algorithm.rs:122]: please use `MAX_PAYLOAD_SIZE` for maximum payload size, we dont want to unpack 10GB from couple of KBs :)

File: Cargo.toml [hunk 4]
```diff
@@ -142,9 +144,9 @@ integration = { path = "core/integration" }
 keyring = { version = "3.6.3", features = ["sync-secret-service", "vendored"] }
 lazy_static = "1.5.0"
 log = "0.4.29"
+lz4 = "1.28.1"
```
Comment (hubcio) [inline Cargo.toml]: maybe use lz4_flex, since its pure rust impl (after you confirm that it works), we'll avoid linking with other libs and our binary size will be smaller for musl builds.

or https://github.com/bozaro/lz4-rs

File: core/sdk/src/clients/consumer.rs [hunk 5]
```diff
@@ -980,6 +977,48 @@ impl Stream for IggyConsumer {
                                 message.payload = Bytes::from(payload);
                                 message.header.payload_length = message.payload.len() as u32;
                             }
+
+                            // maybe decompress
+                            if let Ok(Some(algorithm_value)) = message
+                                .get_user_header(&HeaderKey::from_str("iggy-compression").unwrap())
```
Comment (hubcio) [inline core/sdk/src/clients/consumer.rs]: `pub const COMPRESSION_HEADER_KEY: &str = "iggy-compression";`
and use it in all places

File: core/integration/tests/sdk/producer/compression.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    let client = ClientWrapper::Tcp(TcpClient::create(Arc::new(tcp_client_config)).unwrap());
+    let client = IggyClient::create(client, None, None);
+
+    client.connect().await.unwrap();
+    assert!(client.ping().await.is_ok(), "Failed to ping server");
+
+    login_root(&client).await;
+    init_system(&client).await;
+
+    client.connect().await.unwrap();
```
Comment (hubcio) [inline core/integration/tests/sdk/producer/compression.rs]: no need for second connect

File: core/sdk/src/clients/consumer.rs [hunk 7]
```diff
@@ -980,6 +977,48 @@ impl Stream for IggyConsumer {
                                 message.payload = Bytes::from(payload);
                                 message.header.payload_length = message.payload.len() as u32;
                             }
+
+                            // maybe decompress
+                            if let Ok(Some(algorithm_value)) = message
```
Comment (hubcio) [inline core/sdk/src/clients/consumer.rs]: perhaps put this part into `maybe_decompress`? i'm not insisting, i'd like to hear you opinion about it,

File: core/sdk/src/clients/consumer.rs (hunk 1 above)
Comment (haubur) [reply core/sdk/src/clients/consumer.rs]: Its needed because it returns a Result, no?

```rust 
impl FromStr for CompressionAlgorithm {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "gzip" => Ok(CompressionAlgorithm::Gzip),
            "zstd" => Ok(CompressionAlgorithm::Zstd),
            "lz4" => Ok(CompressionAlgorithm::Lz4),
            "snappy" => Ok(CompressionAlgorithm::Snappy),
            "none" => Ok(CompressionAlgorithm::None),
            _ => Err(format!("Unknown compression type: {s}")),
        }
    }
}
```

File: core/common/src/types/compression/compression_algorithm.rs (hunk 3 above)
Comment (haubur) [reply core/common/src/types/compression/compression_algorithm.rs:122]: good point, done

File: Cargo.toml (hunk 4 above)
Comment (haubur) [reply Cargo.toml]: makes sense, done

File: core/sdk/src/clients/consumer.rs (hunk 5 above)
Comment (haubur) [reply core/sdk/src/clients/consumer.rs]: Missed it. It also was already in the ticket description. Done!

File: core/sdk/src/clients/consumer.rs (hunk 7 above)
Comment (haubur) [reply core/sdk/src/clients/consumer.rs]: That's what I did now. I also liked your codec approach.
However, for decompression the config does not need to be provided as it is read from the user-headers so in order to not unnecessarily instantiate a pseudo or placeholder config in maybe_decompress the maybe_decompress method

```rust
  // core/common/src/types/compression/codec.rs
  pub struct MessageCodec {
      config: ClientCompressionConfig,
  }

  impl MessageCodec {
      pub fn maybe_compress(&self, message: &mut IggyMessage) -> Result<(), IggyError> { ... }
      pub fn maybe_decompress(&self, message: &mut IggyMessage) -> Result<(), IggyError> { ... }
  }
```
would be actually be a static method

```rust
  // core/common/src/types/compression/codec.rs
  pub struct MessageCodec {
      config: ClientCompressionConfig,
  }

  impl MessageCodec {
      pub fn maybe_compress(&self, message: &mut IggyMessage) -> Result<(), IggyError> { ... }
      pub fn maybe_decompress(message: &mut IggyMessage) -> Result<(), IggyError> { ... }
  }
```

So since, the consumer reads inherently and does not need any configuration I left it as is (while having a separate method now).

Still open for other preferences or arguments, though in case I missed something.
